File: src/cpp/tests/test_feline_pk.cpp
#include "doctest.h"
#include "reachability/dynamic/feline_pk.hpp"

using reachability::dynamic::FelinePK;

TEST_CASE("FelinePK: an isolated vertex reaches only itself") {
    FelinePK pk;
    pk.insert_vertex(0);
    pk.insert_vertex(1);
    CHECK(pk.reachable(0, 0));      // reflexive
    CHECK_FALSE(pk.reachable(0, 1));
    CHECK_FALSE(pk.reachable(1, 0));
}

TEST_CASE("FelinePK: remove_vertex drops it from graph and index") {
    FelinePK pk;
    pk.insert_vertex(0);
    pk.insert_vertex(1);
    pk.remove_vertex(1);
    // Alg. 8 must reach all three structures, not just the digraph. Dropping the
    // idx_.remove(v) or r_.erase(v) call — or passing the wrong vertex to either —
    // leaves a live coordinate or a stale set behind, which the graph check alone
    // cannot see.
    CHECK_FALSE(pk.graph().has_vertex(1));
    CHECK_FALSE(pk.index().has(1));
    CHECK_FALSE(pk.rep().contains(1));
    CHECK(pk.reachable(0, 0));
}

TEST_CASE("FelinePK: insert_edge makes the target reachable") {
    FelinePK pk;
    for (unsigned v = 0; v < 3; ++v) pk.insert_vertex(v);
    pk.insert_edge(0, 1);
    pk.insert_edge(1, 2);
    CHECK(pk.reachable(0, 2));
    CHECK_FALSE(pk.reachable(2, 0));
}

TEST_CASE("FelinePK: closing a cycle folds it into one component") {
    FelinePK pk;
    for (unsigned v = 0; v < 3; ++v) pk.insert_vertex(v);
    pk.insert_edge(0, 1);
    pk.insert_edge(1, 2);
    pk.insert_edge(2, 0);           // closes the cycle: {0,1,2} is now one SCC
    CHECK(pk.rep().find(0) == pk.rep().find(1));
    CHECK(pk.rep().find(1) == pk.rep().find(2));

    // No all-pairs loop here: reachable() short-circuits on `a == b` once the
    // representatives coincide, which the two CHECKs above already established. Such a
    // loop would pass over a completely corrupted index — it asserts nothing.
    //
    // What the fold must ALSO do is collapse the three DAG nodes into one. Check the DAG
    // vertex SET, not the edges: fold_cycle strips every edge touching a folded rep before
    // it removes the vertices, so an assertion on dag_succ() being empty holds either way
    // and proves nothing. Membership is what dag_remove_vertex actually changes.
    const auto r = pk.rep().find(0);
    CHECK(pk.graph().dag_out_all().count(r) == 1);    // the representative survives
    for (unsigned v = 0; v < 3; ++v)
        if (v != r) CHECK(pk.graph().dag_out_all().count(v) == 0);   // members left E_DAG
    CHECK(pk.graph().dag_succ(r).count(r) == 0);      // no self-loop was rewired in
}

TEST_CASE("FelinePK: a fold keeps outside reachability intact") {
    FelinePK pk;
    for (unsigned v = 0; v < 5; ++v) pk.insert_vertex(v);
    pk.insert_edge(3, 0);           // 3 feeds the cycle
    pk.insert_edge(0, 1);
    pk.insert_edge(1, 2);
    pk.insert_edge(2, 0);           // fold {0,1,2}
    pk.insert_edge(2, 4);           // the cycle feeds 4
    CHECK(pk.reachable(3, 4));      // through the folded component
    CHECK(pk.reachable(3, 1));
    CHECK_FALSE(pk.reachable(4, 3));
}

TEST_CASE("FelinePK: removing the only link between components cuts reachability") {
    FelinePK pk;
    for (unsigned v = 0; v < 3; ++v) pk.insert_vertex(v);
    pk.insert_edge(0, 1);
    pk.insert_edge(1, 2);
    const int64_t x0 = pk.index().x(0), y0 = pk.index().y(0);
    const int64_t x1 = pk.index().x(1), y1 = pk.index().y(1);
    const int64_t x2 = pk.index().x(2), y2 = pk.index().y(2);
    CHECK(pk.reachable(0, 2));

    pk.remove_edge(1, 2);
    CHECK_FALSE(pk.reachable(0, 2));
    CHECK_FALSE(pk.reachable(1, 2));
    CHECK(pk.reachable(0, 1));

    // Removing an edge never invalidates a topological order: nothing moves.
    CHECK((pk.index().x(0) == x0 && pk.index().y(0) == y0));
    CHECK((pk.index().x(1) == x1 && pk.index().y(1) == y1));
    CHECK((pk.index().x(2) == x2 && pk.index().y(2) == y2));
}

TEST_CASE("FelinePK: a parallel edge keeps the DAG edge alive") {
    // Two distinct E edges between the same pair of components: removing one must NOT
    // remove the E_DAG edge. Removing the second one must.
    FelinePK pk;
    for (unsigned v = 0; v < 4; ++v) pk.insert_vertex(v);
    pk.insert_edge(0, 1); pk.insert_edge(1, 0);   // component {0,1}
    pk.insert_edge(2, 3); pk.insert_edge(3, 2);   // component {2,3}
    pk.insert_edge(0, 2);                         // first link between the components
    pk.insert_edge(1, 3);                         // second link between the SAME components
    CHECK(pk.reachable(0, 3));

    pk.remove_edge(0, 2);
    CHECK(pk.reachable(0, 3));      // still linked through the parallel edge (1,3)
    pk.remove_edge(1, 3);
    CHECK_FALSE(pk.reachable(0, 3)); // unreachable once both links are gone
    CHECK(pk.reachable(0, 1));       // the components themselves are intact
    CHECK(pk.reachable(2, 3));
}

TEST_CASE("FelinePK: removing an edge inside a component splits it") {
    FelinePK pk;
    for (unsigned v = 0; v < 3; ++v) pk.insert_vertex(v);
    pk.insert_edge(0, 1);
    pk.insert_edge(1, 2);
    pk.insert_edge(2, 0);           // fold {0,1,2}
    CHECK(pk.rep().find(0) == pk.rep().find(2));
    for (unsigned v = 0; v < 3; ++v)
        CHECK(pk.index().has(v));   // folded members keep their coordinates

    pk.remove_edge(2, 0);           // the cycle breaks: back to a 0 -> 1 -> 2 chain
    CHECK(pk.rep().find(0) != pk.rep().find(2));
    CHECK(pk.reachable(0, 1));
    CHECK(pk.reachable(1, 2));
    CHECK(pk.reachable(0, 2));
    CHECK_FALSE(pk.reachable(2, 0));
    CHECK_FALSE(pk.reachable(1, 0));
    CHECK_FALSE(pk.reachable(2, 1));
    for (unsigned v = 0; v < 3; ++v)
        CHECK(pk.index().has(v));   // every re-elected representative has a coordinate
}

TEST_CASE("FelinePK: an internal edge whose removal does not split the component") {
    FelinePK pk;
    for (unsigned v = 0; v < 3; ++v) pk.insert_vertex(v);
    pk.insert_edge(0, 1);
    pk.insert_edge(1, 2);
    pk.insert_edge(2, 0);
    pk.insert_edge(1, 0);           // redundant back edge inside the SCC

    pk.remove_edge(1, 0);           // still strongly connected through 1->2->0
    CHECK(pk.reachable(1, 0));
    CHECK(pk.reachable(0, 1));
    CHECK(pk.reachable(2, 1));
    CHECK(pk.reachable(0, 2));
}

TEST_CASE("FelinePK: a split keeps the component's external boundary intact") {
    // 0 -> 1 (external predecessor into the component)
    // SCC {1,2,3}: 1->2, 2->3, 3->1
    // 3 -> 4 (external successor out of the component)
    FelinePK pk;
    for (unsigned v = 0; v < 5; ++v) pk.insert_vertex(v);
    pk.insert_edge(0, 1);
    pk.insert_edge(1, 2); pk.insert_edge(2, 3); pk.insert_edge(3, 1);
    pk.insert_edge(3, 4);
    CHECK(pk.reachable(0, 4));
    CHECK(pk.reachable(3, 1));
    CHECK(pk.reachable(1, 3));

    pk.remove_edge(3, 1);           // breaks the cycle -> chain 1->2->3

    // The chain survives.
    CHECK(pk.reachable(1, 2));
    CHECK(pk.reachable(2, 3));
    CHECK(pk.reachable(1, 3));

    // The cycle is gone.
    CHECK_FALSE(pk.reachable(3, 1));
    CHECK_FALSE(pk.reachable(2, 1));
    CHECK_FALSE(pk.reachable(3, 2));

    // The inbound boundary still works (pins split_component's pred(w) reinsertion loop).
    CHECK(pk.reachable(0, 1));
    CHECK(pk.reachable(0, 2));
    CHECK(pk.reachable(0, 3));
    CHECK(pk.reachable(0, 4));

    // The outbound boundary still works (pins its succ(w) reinsertion loop).
    CHECK(pk.reachable(1, 4));
    CHECK(pk.reachable(2, 4));
    CHECK(pk.reachable(3, 4));

    // Nothing spurious appeared.
    CHECK_FALSE(pk.reachable(4, 0));
    CHECK_FALSE(pk.reachable(4, 3));
    CHECK_FALSE(pk.reachable(1, 0));
}

TEST_CASE("FelinePK: remove_edge no-ops on absent edges and unknown vertices") {
    FelinePK pk;
    pk.insert_vertex(0);
    pk.insert_vertex(1);
    pk.insert_edge(0, 1);

    pk.remove_edge(1, 0);           // this edge never existed
    CHECK(pk.reachable(0, 1));
    pk.remove_edge(5, 6);           // unknown vertices must not crash
    CHECK(pk.reachable(0, 1));

    pk.insert_edge(0, 0);           // self-loop
    pk.remove_edge(0, 0);           // must be a clean no-op
    CHECK(pk.reachable(0, 0));
    CHECK(pk.reachable(0, 1));      // the real edge was left alone

    pk.remove_edge(0, 1);           // the real removal still works afterwards
    CHECK_FALSE(pk.reachable(0, 1));
}

File: tests/test_feline_pk.py
import numpy as np
import pytest
from hypothesis import given, settings, strategies as st

from pyreachability import Graph, catalog
from pyreachability._oracle import reachable_set
from pyreachability.dynamic import FelinePK
from pyreachability.dynamic import base as dyn_base


def _oracle(n, edges, u, v):
    return v in reachable_set(n, edges)[u]


def test_seeds_from_a_static_graph_then_evolves():
    g = Graph.from_edges(np.array([0, 1], np.int32), np.array([1, 2], np.int32), num_nodes=3)
    idx = FelinePK()
    idx.build(g)
    assert idx.query(0, 2)
    idx.remove_edge(1, 2)
    assert not idx.query(0, 2)
    idx.insert_edge(0, 2)
    assert idx.query(0, 2)


def test_declares_all_four_abilities():
    assert FelinePK.supports == frozenset({
        dyn_base.EDGE_INSERT, dyn_base.EDGE_DELETE,
        dyn_base.VERTEX_INSERT, dyn_base.VERTEX_DELETE,
    })
    # Batch insertion is Feline-PK's declared future work: it IS the loop today.
    assert dyn_base.EDGE_INSERT_BATCH not in FelinePK.supports


def test_is_registered_as_dynamic():
    assert "feline-pk" in catalog.methods(kind="dynamic")
    assert "feline-pk" not in catalog.methods(kind="static")


def test_query_before_build_raises():
    idx = FelinePK()
    with pytest.raises(RuntimeError):
        idx.query(0, 1)
    with pytest.raises(RuntimeError):
        idx.query_batch(np.array([[0, 1]], np.int32))


def test_scc_fold_happens_during_build():
    """build() seeds edge-by-edge into the same incremental insert_edge as
    runtime updates use, so a cycle present at seed time must fold during
    build() itself -- there is no separate one-shot condensation pass here
    (unlike the static half). Seed a 3-cycle with a tail on each end.

    Edge order inside build() matters here, not just for style: it iterates
    vertices in ascending id and inserts each one's out-edges in turn, so an
    edge whose source has a *smaller* id than the cycle-closing edge's source
    lands in the DAG before the fold happens -- landing 0->1 and 2->4 before
    the closing edge 3->1 forces the fold's own I/O predecessor/successor
    rewiring to run with both sets non-empty. Get this backwards (e.g. route
    the tail through vertex 3, whose out-edges are processed last) and the
    tail edges get added as fresh, ordinary edges *after* the fold instead of
    exercising the fold's I/O rewiring at all -- silently making this test
    pass on a build that drops that rewiring. Confirmed by mutation: dropping
    the I-side rewiring line in fold_cycle breaks 0->1 in this exact layout
    while leaving every other assertion in this file green."""
    # 0 -> {1,2,3} (a 3-cycle, closed by the last edge 3->1) -> 4
    g = Graph.from_edges(
        np.array([0, 1, 2, 2, 3], np.int32),
        np.array([1, 2, 3, 4, 1], np.int32),
        num_nodes=5,
    )
    idx = FelinePK()
    idx.build(g)
    cycle = [1, 2, 3]
    for a in cycle:
        for b in cycle:
            assert idx.query(a, b), f"{a}->{b} should hold: same SCC"
    for a in cycle:
        assert idx.query(0, a)
        assert not idx.query(a, 0)
        assert idx.query(a, 4)
        assert not idx.query(4, a)
    assert not idx.query(4, 0)


def test_remove_edge_splits_a_component_all_four_directions():
    """Two 2-cycles {0,1} and {2,3}, joined into one 4-cycle by 1->2 and 3->0
    so build() folds all four into a single SCC. Vertex 4 has a direct edge
    into the component (4->0) and vertex 5 receives a direct edge out of it
    (3->5) -- neither 4 nor 5 is itself folded into the SCC, so after the
    split their DAG links can only be restored by split_component's re-insert
    loop scanning the *other* endpoint (4->0 needs pred(0), 3->5 needs
    succ(3); an edge fully internal to the split component, like the
    surviving bridge 3->0, would be reinstated by either loop alone and so
    would not tell the two apart).

    Removing 1->2 breaks only the forward bridge: {0,1} and {2,3} stay
    individually cyclic, and the surviving 3->0 bridge makes {2,3} -> {0,1}
    reachable but not the reverse. Checks all four directions: within-A,
    within-B, B->A, A->B, plus the two external edges crossing the split."""
    g = Graph.from_edges(
        np.array([0, 1, 1, 2, 3, 3, 4, 3], np.int32),
        np.array([1, 0, 2, 3, 2, 0, 0, 5], np.int32),
        num_nodes=6,
    )
    idx = FelinePK()
    idx.build(g)
    # Before removal: one SCC of all four, 4 reaches everything, everything reaches 5.
    for a in range(4):
        for b in range(4):
            assert idx.query(a, b), f"pre-split {a}->{b} should hold: one SCC"
    for a in range(4):
        assert idx.query(4, a)
        assert idx.query(a, 5)
    assert idx.query(4, 5)

    idx.remove_edge(1, 2)

    # within {0, 1}: untouched 2-cycle
    assert idx.query(0, 1)
    assert idx.query(1, 0)
    # within {2, 3}: untouched 2-cycle
    assert idx.query(2, 3)
    assert idx.query(3, 2)
    # {2, 3} -> {0, 1}: surviving bridge 3->0
    assert idx.query(2, 0)
    assert idx.query(2, 1)
    assert idx.query(3, 0)
    assert idx.query(3, 1)
    # {0, 1} -> {2, 3}: bridge removed, no path back
    assert not idx.query(0, 2)
    assert not idx.query(0, 3)
    assert not idx.query(1, 2)
    assert not idx.query(1, 3)
    # external edge INTO the split component (4->0): direct edge, must survive
    # regardless of the bridge -- exercises split_component's pred(w) re-insert.
    assert idx.query(4, 0)
    assert idx.query(4, 1)
    assert not idx.query(4, 2)   # only reachable by crossing the now-removed bridge
    assert not idx.query(4, 3)
    # external edge OUT of the split component (3->5): exercises succ(w) re-insert.
    assert idx.query(2, 5)
    assert idx.query(3, 5)
    assert not idx.query(0, 5)   # A can no longer reach B, so cannot reach 5 either
    assert not idx.query(1, 5)


def test_query_batch_matches_query_loop():
    g = Graph.from_edges(np.array([0, 1, 3], np.int32), np.array([1, 2, 4], np.int32), num_nodes=5)
    idx = FelinePK()
    idx.build(g)
    idx.remove_edge(1, 2)
    idx.insert_edge(2, 3)
    pairs = [(a, b) for a in range(5) for b in range(5)]
    batch = idx.query_batch(np.array(pairs, dtype=np.int32))
    loop = [idx.query(a, b) for a, b in pairs]
    assert list(batch) == loop
    assert batch.dtype == np.bool_


def test_query_batch_has_no_upper_bound_and_tolerates_bootstrapped_vertices():
    """query_batch deliberately has no upper-bound check: Feline-PK has no
    fixed vertex count, and insert_edge bootstraps vertices past build()'s
    original n. An id the index has never seen at all must be tolerated too
    (reachable() returns False), matching query()'s behaviour -- not raise."""
    g = Graph.from_edges(np.array([0], np.int32), np.array([1], np.int32), num_nodes=2)
    idx = FelinePK()
    idx.build(g)
    idx.insert_edge(1, 50)  # bootstraps vertex 50, never part of build()
    assert idx.query(0, 50) is True
    assert idx.query(0, 999) is False   # 999 was never created anywhere

    pairs = np.array([[0, 50], [50, 0], [0, 999]], dtype=np.int32)
    batch = idx.query_batch(pairs)
    assert list(batch) == [idx.query(0, 50), idx.query(50, 0), idx.query(0, 999)]
    assert list(batch) == [True, False, False]


def test_negative_vertex_id_is_rejected():
    g = Graph.from_edges(np.array([0], np.int32), np.array([1], np.int32), num_nodes=2)
    idx = FelinePK()
    idx.build(g)
    with pytest.raises(IndexError):
        idx.query(-1, 0)
    with pytest.raises(IndexError):
        idx.insert_edge(-1, 0)
    with pytest.raises(IndexError):
        idx.remove_edge(0, -1)
    with pytest.raises(IndexError):
        idx.insert_vertex(-1)
    with pytest.raises(IndexError):
        idx.remove_vertex(-1)
    with pytest.raises(IndexError):
        idx.query_batch(np.array([[0, -1]], dtype=np.int32))


@settings(max_examples=50, deadline=None)
@given(
    n=st.integers(min_value=2, max_value=6),
    ops=st.lists(
        st.tuples(st.booleans(), st.integers(0, 5), st.integers(0, 5)),
        min_size=1, max_size=25,
    ),
)
def test_update_sequences_agree_with_the_oracle(n, ops):
    """The oracle with a time axis: re-check every pair after every update."""
    idx = FelinePK()
    idx.build(Graph.from_edges(np.empty(0, np.int32), np.empty(0, np.int32), num_nodes=n))
    edges = []
    for inserting, u, v in ops:
        if u >= n or v >= n or u == v:
            continue
        if inserting:
            idx.insert_edge(u, v)
            if (u, v) not in edges:
                edges.append((u, v))
        else:
            idx.remove_edge(u, v)
            edges = [e for e in edges if e != (u, v)]
        for a in range(n):
            for b in range(n):
                assert idx.query(a, b) == _oracle(n, edges, a, b), (
                    f"after {'insert' if inserting else 'remove'} {u}->{v}: {a}->{b}"
                )


@settings(max_examples=30, deadline=None)
@given(
    n=st.integers(min_value=2, max_value=6),
    ops=st.lists(
        st.tuples(st.booleans(), st.integers(0, 5), st.integers(0, 5)),
        min_size=1, max_size=15,
    ),
)
def test_query_batch_agrees_with_query_through_updates(n, ops):
    """query_batch's own marshalling (dtype casts, reshape, min-check) is a
    second crossing of the Python/C++ boundary that scalar query() does not
    exercise -- check it stays in lockstep with query() through the same
    update sequences the time-axis oracle uses."""
    idx = FelinePK()
    idx.build(Graph.from_edges(np.empty(0, np.int32), np.empty(0, np.int32), num_nodes=n))
    for inserting, u, v in ops:
        if u >= n or v >= n or u == v:
            continue
        if inserting:
            idx.insert_edge(u, v)
        else:
            idx.remove_edge(u, v)
    pairs = [(a, b) for a in range(n) for b in range(n)]
    batch = idx.query_batch(np.array(pairs, dtype=np.int32))
    loop = [idx.query(a, b) for a, b in pairs]
    assert list(batch) == loop
